app/scripts/lib/channels/duplex.js

```diff
@@ -0,0 +1,185 @@
+    },
+
+    remove: function (messageId) {
+      var outstanding = this.get(messageId);
+      if (outstanding) {
+        this._window.clearTimeout(outstanding.timeout);
+        delete this._requests[outstanding.messageId];
+      }
+    },
+
+    get: function (messageId) {
+      return this._requests[messageId];
+    },
+
+    clear: function () {
+      for (var messageId in this._requests) {
+        this.remove(this._requests[messageId]);
+      }
+    }
+  };
+
+  function DuplexChannel() {
+  }
+
+  _.extend(DuplexChannel.prototype, new BaseChannel(), {
+    promise: true,
+    init: function (options) {
+      this._sender = options.sender;
+      this._receiver = options.receiver;
+      this._receiver.on('message', this.onMessageReceived.bind(this));
+
+      this._outstandingRequests = new OutstandingRequests({
+        window: options.window,
+        sendTimeoutLength: options.sendTimeoutLength
+      });
+    },
+
+    teardown: function () {
+      this._outstandingRequests.clear();
+      this._sender.teardown();
+      this._receiver.teardown();
+    },
+
+    /**
+     * Send a message, do not expect a response.
+     *
+     * @param {String} command
+     * @param {Object} [data]
+     * @return {Promise}
+     *        Promise will resolve whenever message is sent.
+     */
+    send: function (command, data) {
+      var self = this;
+      return p()
+        .then(function () {
+          return self._sender.send(command, data, null);
+        });
+    },
+
+    /**
+     * Send a message, expect a response.
+     *
+     * @param {String} command
+     * @param {Object} [data]
+     * @return {Promise}
+     *        Promise will resolve when the response is received.
+     */
+    request: function (command, data) {
+      var self = this;
+
+      var messageId = this.createMessageId(command, data);
+      var outstanding = {
+        deferred: p.defer(),
+        command: command,
+        data: data,
+        messageId: messageId
+      };
+
+      // save the data beforehand in case the response is synchronous.
+      self._outstandingRequests.add(messageId, outstanding);
+
+      return p()
+        .then(function () {
+          return self._sender.send(command, data, messageId);
+        })
+        .then(function () {
+          return outstanding.deferred.promise;
+        })
+        .fail(function (err) {
+          // If there was a problem sending the request is
+          // no longer outstanding.
+          self._outstandingRequests.remove(messageId);
+          throw err;
+        });
+    },
+
+    /**
+     * Create a messageId for a given command/data combination.
+     *
+     * messageId is sent to the relier who is expected to respond
+     * with the same messageId. Used to keep track of outstanding requests.
+     *
+     * @param {String} command
+     * @param {Object} [data]
+     * @return {String}
+     */
+    createMessageId: function (command, data) { //jshint ignore: line
```

Comment:
createMessageId and parseMessage act in tandem (one to send, one to receive) and are functions so they can be overridden by future subclasses whose remote side may not support echoing back a `messageId`.